lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,439 @@
+        Args:
+            key: Header parameters key.
+            value: Value to be added for the key.
+        """
+        self.headers[key] = value
+
+    def del_param_headers(self, key: str):
+        """
+        Remove a specific header parameter.
+        Args:
+            Key to delete from the headers.
+        """
+        del self.headers[key]
+
+    def set_well_known_endpoints(self):
+        """
+        Sets the well-known configuration endpoints for the current OIDC
+        client. This includes common useful endpoints for decoding tokens,
+        getting user information etc.
+        ref: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
+        """
+        well_known_endpoint = "/.well-known/openid-configuration"
+        well_known_uri = f"{self.server_url}{self.realm_name}{well_known_endpoint}"
+        endpoints_json = self._get(well_known_uri).json()
+        try:
+            OpenIDClient.AUTHORIZATION_ENDPOINT = endpoints_json[
+                "authorization_endpoint"
+            ]
+            OpenIDClient.TOKEN_ENDPOINT = endpoints_json["token_endpoint"]
+            OpenIDClient.USERINFO_ENDPOINT = endpoints_json["userinfo_endpoint"]
+            OpenIDClient.REVOCATION_ENDPOINT = endpoints_json["revocation_endpoint"]
+            OpenIDClient.JWKS_ENDPOINT = endpoints_json["jwks_uri"]
+        except KeyError as e:
+            self.logger.exception(
+                "Missing endpoint {!r} at {}; Endpoints found: {}",
+                str(e),
+                well_known_uri,
+                endpoints_json,
+            )
+            raise
+
+    def get_oidc_public_key(self, token: str):
+        """
+        Returns the OIDC client public key that can be used for decoding
+        tokens offline.
+        Args:
+            token: Third party token to extract the signing key
+        Returns:
+            OIDC client public key
+        """
+        jwks_client = jwt.PyJWKClient(self.JWKS_URI)
+        pubkey = jwks_client.get_signing_key_from_jwt(token).key
+        return pubkey
+
+    def get_user_token(
+        self,
+        username: str,
+        password: str = None,
+        grant_type: str = "password",
+        scope: Union[str, List[str]] = "openid profile email",
+        **extra,
+    ) -> JSON:
+        """
+        The token endpoint is used to obtain tokens. Tokens can either be
+        obtained by exchanging an authorization code or by supplying credentials
+        directly depending on what authentication flow is used. The token
+        endpoint is also used to obtain new access tokens when they expire.
+        http://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint
+        Note: Some OIDC clients only accept authorization_code as a grant_type
+        for getting a token, getting a token directly from these clients with
+        username and password over an API call will result in Forbidden error.
```

Comment:
But I think you are thinking only when there is a client involved like a dashboard or something else. I was thinking of the users who would not use any client and just rely on CLI to login and get tokens to access the Pbench resources. 

Now, we have already decided to use the third-party IDP token and to generate and revoke/logout those tokens of course the user would ideally have to use a dashboard-like client. However, alongside the third-party logins if we decide to have Pbench-specific user registration (with our backend database enabled with Keycloak or LDAP) then we would definitely need these functionalities for the CLI users to generate tokens. I am not sure how the registration page is gonna look like at this point but _I_ _think_ some of the Pbench users are heavy CLI users and they never go to the dashboard and we have to provide a way for these users to quickly register and get the tokens for various needs. So I am inclined to keep it for now and then remove it later if we don't need it (since its likely that we might need it). @dbutenhof any thought?

Reply:
We definitely need to think about how much of the server we want to expose by "CLI". Right now we have `pbench-generate-token` and `pbench-results-move`: there's been some discussion about whether those use "special" tokens that might not even be from Keycloak. This could include GitHub-like API tokens that would be long-lasting... but which we still might want to be able to revoke.

On the other hand, even if we go this way, it's not clear *this* API will be exactly what we need, and maybe it's better to be agile and not to try to anticipate where we'll end up.

Reply:
For support for CLI functions like `pbench-results-move` we're going to need a different kind of token...I hesitate to even call it a "token", because it may well be generated and evaluated using a completely different mechanism.  This problem has already been solved by GitHub, AWS, Quay, COPR, and others:  you log into a browser-based application or equivalent which performs the dance with your IDP of choice, and then the server issues you an...let's call it an "access key"...which you can then use with the CLI and similar requests.  The access key has an arbitrary lifetime, and the Pbench Server can validate it when it comes in with a request.  If the user wants, they can invalidate/delete the key using the same sort of protocol used to create it.

The point is, the creation (and deletion) of an access key requires a "login" via the conventional means.  However, the use of an access key does _not_ require a login.

And so, the Pbench Server never accesses the IDP (except for on-line validation, which, perhaps, is never necessary, if the IDP tokens are short-enough lived), and so access keys can be implemented in whatever way we see fit.  (We might want to utilize Keycloak for generating them, or we might not.)

As for `pbench-generate-token`, I think that that will likely be obsolete.  Unless we make it pretty fancy, it won't be able to do the dance with the IDM.  The best it will be able to do is to point the user at a browser to do it, and that'll be so cumbersome that we might as well just go with the access key approach.  (Either way, the Pbench Server doesn't need to be involved in the authentication part of things; just the access key management stuff.)

> if we decide to have Pbench-specific user registration (with our backend database enabled with Keycloak or LDAP) then we would definitely need these functionalities for the CLI users to generate tokens.

Why?  Why can't we just have the Pbench Server point the Dashboard/client to our private Keycloak server (or, configure our Keycloak broker to work with our private Keycloak server)?  That way, we get to have private, Pbench-specific users, but the Pbench-server isn't involved in the process.

Nevertheless, as Dave suggests, let's try to implement only what we need.  We want open, flexible designs, but we want minimal, targeted implementations.

Reply:
Yeah, okay we might as well have to remove all the user registration and login stuff from the server and coordinate that with the dashboard. If we do remove all the user registration and login from the server then we need to have our Keycloak broker instance up and running somewhere where the dashboard can point to in the meantime. But let's cross that bridge when we get there (hopefully pretty soon).

Reply:
Yes; once you have the basic Keycloak integration up and running, we'll need to support dashboard login and server authentication before we can deploy the new server even for staging (or individual testing). That includes [PBENCH-839](https://issues.redhat.com/browse/PBENCH-838) to expose the server's Keycloak link to the dashboard code... and I think [PBENCH-535](https://issues.redhat.com/browse/PBENCH-535) can be interpreted as implementation of the authentication redirect dance given the appropriate server and Keycloak configuration.

Reply:
>  If we do remove all the user registration and login from the server then we need to have our Keycloak broker instance up and running somewhere where the dashboard can point to in the meantime.

In the context of this PR, I'm not talking about the existing login and token implementation in the Pbench Server.  That can stay around as long as it is useful and doesn't get in the way.  The code changes presented in this PR are for new code, and, really, just foundations for new code.  We will build or extend Pbench Server APIs based on this code.  Once those are up (and the IDP broker is deployed), then we can consider whether to switch over from our old tokens to the Keycloak-based ones or whether we want to (temporarily) wire the Pbench Server to accept either one.  So, we should be able to lay a smooth path forward (at least for development and testing).

I am **_not_** advocating tearing out the existing login mechanism in this PR.  I'm suggesting that we decrease the amount of code that we are _adding_ with this PR.

Reply:
Yep definitely not trying to tear apart the existing login mechanism in this PR. We need at least until we have the IDP broker deployed.